File: src/ts/examples/read-local.ts
/** Whole file (or one bbox) out as CityJSONSeq, on stdout.
 *
 *      node examples/read-local.ts in.fcb > out.city.jsonl
 *      node examples/read-local.ts in.fcb 84500 445800 85000 446500
 *
 *  With a bbox the R-tree answers first and only the matching features
 *  are read; without one this is a straight sequential scan. Progress
 *  goes to stderr so stdout stays a clean CityJSONSeq stream.
 */
import { toCityJSONFeature, toCityJSONMetadata } from '@cityjson/flatcitybuf'
import { fromFile } from '@cityjson/flatcitybuf/node'

const [path, ...rest] = process.argv.slice(2)
if (path === undefined || (rest.length !== 0 && rest.length !== 4)) {
  console.log('usage: node examples/read-local.ts <file.fcb> [minX minY maxX maxY]')
  process.exit(2)
}

const reader = await fromFile(path)
try {
  const info = reader.header.info
  const crs = info.referenceSystem === undefined ? '' : `, ${info.referenceSystem}`
  console.error(`${info.featuresCount} features, CityJSON ${info.version}${crs}`)

  // Line 0 is the CityJSON header: transform, metadata, and the geometry
  // templates and appearance palette when the file has them.
  console.log(JSON.stringify(toCityJSONMetadata(reader.header)))

  const cursor =
    rest.length === 4
      ? await reader.select({
          spatial: {
            kind: 'bbox',
            value: rest.map(Number) as [number, number, number, number],
          },
        })
      : await reader.selectAll()

  // featuresCount is the TOTAL number of matches, not the page size --
  // it is unaffected by limit/offset.
  if (rest.length === 4) console.error(`${cursor.featuresCount} feature(s) in the bbox`)

  let n = 0
  for await (const feature of cursor) {
    console.log(JSON.stringify(toCityJSONFeature(feature, reader.header)))
    n += 1
  }
  console.error(`wrote ${n} feature(s)`)
} finally {
  await reader.close()
}

File: src/py/examples/read_local.py
"""Whole file (or one bbox) out as CityJSONSeq, on stdout.

    python examples/read_local.py in.fcb > out.city.jsonl
    python examples/read_local.py in.fcb 84500 445800 85000 446500

With a bbox the R-tree answers first and only the matching features are
decoded; without one this is a straight sequential scan. Progress goes
to stderr so stdout stays a clean CityJSONSeq stream.
"""

from __future__ import annotations

import json
import sys

import flatcitybuf as fcb


def main(argv: list[str]) -> int:
    path = argv[0]
    box = tuple(float(v) for v in argv[1:5]) if len(argv) == 5 else None

    reader = fcb.FcbReader.open_file(path)
    info = reader.header.info
    print(
        f"{info.features_count} features, CityJSON {info.cityjson_version}"
        f"{', ' + info.crs if info.crs else ''}",
        file=sys.stderr,
    )

    # Line 0 is the CityJSON header: transform, metadata, and the
    # geometry templates and appearance palette if the file has them.
    print(json.dumps(fcb.to_cityjson_metadata(reader.header)))

    if box is None:
        features = reader.select_all()
    else:
        # search_rtree is a free function over a RangeReader and hands
        # back byte offsets, not features; feature_at is the way back.
        hits = fcb.search_rtree(
            reader.range_reader,
            reader.header.layout.rtree_begin,
            info.features_count,
            info.index_node_size,
            box,  # type: ignore[arg-type]
        )
        print(f"{len(hits)} feature(s) in the bbox", file=sys.stderr)
        features = (reader.feature_at(h) for h in hits)

    n = 0
    for feature in features:
        print(json.dumps(fcb.to_cityjson_feature(feature, reader.header)))
        n += 1
    print(f"wrote {n} feature(s)", file=sys.stderr)
    return 0


if __name__ == "__main__":
    if len(sys.argv) not in (2, 6):
        print(__doc__)
        raise SystemExit(2)
    raise SystemExit(main(sys.argv[1:]))
